windows/start.sh

```diff
@@ -56,7 +56,32 @@ cat << EOF
 EOF
 echo -e "${BLUE}docker${NC} is configured to use the ${GREEN}$VM${NC} machine with IP ${GREEN}$($DOCKER_MACHINE ip $VM)${NC}"
 echo "For help getting started, check out the docs at https://docs.docker.com"
-echo "NOTE: When using interactive commands, prepend winpty. Examples: 'winpty docker run -it ...', 'winpty docker exec -it ...'."
+
+# Fix tty access with winpty
+if [[ -z "$docker_bin" ]] ; then
+  docker_bin=$(which docker)
+fi
+function docker {
+  if [[ -z ${docker_old_IFS+x} ]] ; then
+    docker_old_IFS=$IFS
+  fi
+ IFS=''
+ docker_use_winpty=0
+ while read -r line; do
+   echo $line
+    if [[ $line == "cannot enable tty mode on non tty input" ]] ; then
+      docker_use_winpty=1
+    fi;
+ done < <("$docker_bin" $@ 2>&1)
+ if [[ $docker_use_winpty == 1 ]] ; then
+   echo "Using winpty"
+   winpty $docker_bin $@
```

Comment:
> Doesn't this flow mean that we'll potentially run whatever command the user specifies more than once? Or does the "cannot enable tty mode on non tty input" message fatal out and prevent Docker from running at all?

Yes, the "cannot enable tty mode ..." fatals out and the command is never run, with the fix, it is then run successfully only once.

> The 2>&1 means that if a user's shell does support proper ttys, their output will have stdout and stderr smashed together instead of separated, which I think would be a surprise. 

Yes, the 2>&1 get's smashed together, but this fix is only in the start.sh from Windows, which for certain doesn't support proper ttys (al least not with mintty, which is what is used by Git bash). The start.sh from Mac keeps unmodified. This bug only affects mintty (Windows) terminals.

> Do you think we could use a simpler method of detecting whether we need winpty?

Another possibility I haven't tested would be to use exit codes, but I'm not sure how confident I can be that Docker / docker-machine will always output the same exit code in every Windows flavor, in every future version of Git Bash (even if they change again their default terminal emulator), etc. I actually trust more the output message.

> What are the situations that cause it to come up? Any output, or just specific use cases? 

The situations that raise the tty issue are when someone tries to get inside the container in an interactive way with a tty, I think, when setting the ```-t``` flag or similar, as in:

```docker exec -it supercontainer bash```

or 

```docker run -it debian bash```

or

``` docker run -it debian apt-get update && apt-get install -y python && python```

Also, if the user tries to run "sh", "zsh" or probably other unknown interactive shells or similar interactive programs. 

(but remember, this a Windows only issue and Windows only fix).

> Could we detect whether it's necessary once, and only define the function overloading docker when it is necessary?

I think the possibilities of running into the issue are quite many, I think it is easier to catch the error than trying to overload ```docker``` and take into account every possibility (from an infinite set). Also, always running docker with winpty wouldn't be as cool and would raise other issues. You have already done that and then rolled back.

---

I think this is the best approach that can be taken in the short-term. Maybe the correct solution would be to fix mintty (but I think no one will do that soon). And I think that this solution facilitates the most starting with- and using Docker, especially for newcomers, which, I think, is the main point of Docker Toolbox. It doesn't solve every possible related issue, but it solves most of the current (and sort of critical) ones.

Also, I know in the previous 3 or 5 versions you tried to solve this by SSH-ing into the VM which turned out to be confusing / inconvenient for some users, then you tried creating an alias of ```docker``` with ```winpty docker``` that raised other issues, and in the latest version there's just a message saying "NOTE: When using interactive commands, prepend winpty...". I think you may try this one and it will probably work for many people.

---

A possible issue that may happen is if someone tries to run an interactive command while at the same time uses slashes that get changed by winpty, as documented in this comment: https://github.com/docker/toolbox/issues/230#issuecomment-147122459 by @gmanfunky, something like:

```docker run -it google/debian bash```

although this would work:

```docker run -d google/debian ping google.com```

because it doesn't try to run interactively with a tty.

And this would also work:

```docker run -it debian bash```

because it doesn't have slashes.

But I think that those are more like corner cases, because there must be the combination of an interactive command, as running bash, and a slash in the full command.

Also, with the actual help message, the person would try that by hand (prepending winpty) and end up with the same problem.

---

And by the way, I'm basing the fix in my babun-docker project, so this way of doing it has already gone though and solved the issues raised by several other ways to do something similar:

* Catching the full output and then finding the string before outputting it: that would hold the output until the process ends, impeding the user to know the current state of that process, which would be terrible in a long operation as a ```docker pull```.
* Printing the output directly to the screen: that would break redirection and would break commands as ```docker stop $(docker ps -q)```.
* Using always winpty: that would break the commands with slashes when they could have been used, for example: ```docker pull google/debian```

---

Also, you can test it personally (without building again) by copying this "```start.sh```" file into your "```C:\Program Files\Docker Toolbox```" directory (assuming you have a Windows at hand to test).